memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -109,55 +110,41 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 
-	// Check if the Deployment already exists, if not create a new one
-	deployment := &appsv1.Deployment{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: memcached.Name, Namespace: memcached.Namespace}, deployment)
-	if err != nil && errors.IsNotFound(err) {
-		// Define a new Deployment
-		dep := r.deploymentForMemcached(memcached)
-		reqLogger.Info("Creating a new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-		err = r.client.Create(context.TODO(), dep)
-		if err != nil {
-			reqLogger.Error(err, "Failed to create new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-			return reconcile.Result{}, err
-		}
-		// Deployment created successfully - return and requeue
-		// NOTE: that the requeue is made with the purpose to provide the deployment object for the next step to ensure the deployment size is the same as the spec.
-		// Also, you could GET the deployment object again instead of requeue if you wish. See more over it here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler
-		return reconcile.Result{Requeue: true}, nil
-	} else if err != nil {
-		reqLogger.Error(err, "Failed to get Deployment.")
-		return reconcile.Result{}, err
-	}
+	// Update Deployment if it has changed to something other than expected. If the Deployment
+	// doesn't exist it will be created (initial creation is not really a special case).
+	deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: memcached.Name, Namespace: memcached.Namespace}}
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, deployment, func(ignore runtime.Object) error {
+		// Set/reset deployment to expected state
+		deploymentForMemcached(memcached, reqLogger, deployment)
 
-	// Ensure the deployment size is the same as the spec
-	size := memcached.Spec.Size
-	if *deployment.Spec.Replicas != size {
-		deployment.Spec.Replicas = &size
-		err = r.client.Update(context.TODO(), deployment)
-		if err != nil {
-			reqLogger.Error(err, "Failed to update Deployment.", "Deployment.Namespace", deployment.Namespace, "Deployment.Name", deployment.Name)
-			return reconcile.Result{}, err
-		}
+		// Set Memcached instance as the owner of the Deployment.
+		// If error is returned, it will be passed up to outer CreateOrUpdate function
+		return controllerutil.SetControllerReference(memcached, deployment, r.scheme)
```

Comment:
IMHO the owner ref should be set only when the object is created.  Also, it can lead to the users the idea they should use always the `controllerutil.SetControllerReference` in the updates. However, they can decide update objects which should not be owned by the Controller. Is it make sense?